### PR TITLE
Enable Custom Lowering for fabs.v8f16 on AVX

### PR DESCRIPTION
**Pull request by david-xl**
[X86]: Enable custom lowering for fabs.v8f16 on AVX

Currently, custom lowering of fabs.v8f16 requires AVX512FP16, which is too restrictive. For v8f16 fabs lowering, no instructions in AVX512FP16 are needed.  Without the fix, horribly inefficient code is generated without AVX512FP16. Note instcombiner generates calls to intrinsics @llvm.fabs.v8f16 when simplifyping AND <8 x half> operations.

### CODE REVIEW

**Comment by david-xl** [comment]
This fix improves Eigen pcos<Packet8h> performance by ~65% on Haswell.

**Review by RKSimon — COMMENTED** [review]
@david-xl - please can you rebase this?

**Comment by RKSimon** [comment]
This looks like the merge has gone wrong?

**Comment by david-xl** [comment]
> This looks like the merge has gone wrong?

Right -- I messed it up (new to the workflow). Will fix.

**Comment by david-xl** [comment]
> This looks like the merge has gone wrong?

Fixed the bad merge. PTAL.

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 1]
```diff
@@ -2238,6 +2238,9 @@ X86TargetLowering::X86TargetLowering(const X86TargetMachine &TM,
     }
   }
 
+  if (!Subtarget.useSoftFloat() && Subtarget.hasAVX())
+    setOperationAction(ISD::FABS, MVT::v8f16, Custom);
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:2242]: Please can you move this inside the main "Subtarget.hasAVX()" section, around line 1425?

File: llvm/test/CodeGen/X86/vec_fabs.ll [hunk 2]
```diff
@@ -137,6 +137,45 @@ define <4 x double> @fabs_v4f64(<4 x double> %p) {
...
+; X64-AVX2:       # %bb.0:
+; X64-AVX2-NEXT:    vpbroadcastw {{\.?LCPI[0-9]+_[0-9]+}}(%rip), %xmm0
+; X64-AVX2-NEXT:    vpand (%rdi), %xmm0, %xmm0
+; X64-AVX2-NEXT:    retq
+
+  %v = load <8 x half>, ptr %p, align 16
+  %nnv = call <8 x half> @llvm.fabs.v8f16(<8 x half> %v)
+  ret <8 x half> %nnv
+}
+declare <8 x half> @llvm.fabs.v8f16(<8 x half> %p)
```
Comment (RKSimon) [inline llvm/test/CodeGen/X86/vec_fabs.ll:177]: Please can you add v16f16 / v32f16 test coverage as well?

**Comment by david-xl** [comment]
Addressed review comments. PTAL.

File: llvm/test/CodeGen/X86/vec_fabs.ll [hunk 3]
```diff
@@ -137,6 +139,86 @@ define <4 x double> @fabs_v4f64(<4 x double> %p) {
...
+; X64-AVX512FP16-LABEL: fabs_v16f16:
+; X64-AVX512FP16:       # %bb.0:
+; X64-AVX512FP16-NEXT:  vpbroadcastw {{\.?LCPI[0-9]+_[0-9]+}}(%rip), [[YMM:%ymm[0-9]+]]
+; X64-AVX512FP16-NEXT:  vpand (%rdi), [[YMM]], [[YMM]]
+; X64-AVX512FP16-NEXT:  retq
+; 
+  %v = load <16 x half>, ptr %p, align 32
+  %nnv = call <16 x half> @llvm.fabs.v16f16(<16 x half> %v)
+  ret <16 x half> %nnv
+}
```
Comment (RKSimon) [inline llvm/test/CodeGen/X86/vec_fabs.ll:198]: Just let the script generate the codegen, even if its really poor for pre-AVX512FP16

**Review by RKSimon — APPROVED** [review]
LGTM with one minor

File: llvm/test/CodeGen/X86/vec_fabs.ll (hunk 3 above)
Comment (david-xl) [reply llvm/test/CodeGen/X86/vec_fabs.ll:198]: Ok. I will have a followup patch to update the test.

Comment (david-xl) [reply llvm/test/CodeGen/X86/vec_fabs.ll:198]: thanks. Looks like you have already updated and cleaned up the test.

Comment (RKSimon) [reply llvm/test/CodeGen/X86/vec_fabs.ll:198]: Yup, we're still doing a really bad job with v16f16/v32f16 on AVX targets - please can you take a look?

Comment (david-xl) [reply llvm/test/CodeGen/X86/vec_fabs.ll:198]: Will take a look.

**Pull request merged**